### ISSUE TITLE
Sails 0.11 loads very slowly with many models

### ISSUE THREAD

**Issue by bradshja** [balderdashy/sails#2669]
We have a Sails based API with 786 each of models and controllers.  In Sails 10.x it loaded in a few seconds, but upgrading to 0.11 causes it to take over 2 minutes to load.

I eventually narrowed it down to commit 1fdcc859eead5bbfa735f2887d0c7d3498832314.  The addSailsReference (renamed to bindToSails later?) function is the culprit, and if you change it to return the callback rather than loop through the passed in modules like it does, then Sails loads in a few seconds again.

I don't have any fixes that maintain the old functionality, so no PR for this.

**Comment by tjwebb** [comment]
Try setting `dynamicFinders: false` in config/models.js. It uses less RAM and will boot up a little bit faster.

**Comment by bradshja** [comment]
I just tried that, and if it had any effect it wasn't noticeable.

**Comment by tjwebb** [comment]
How many models do you have?

**Comment by bradshja** [comment]
786.  I understand that's probably far more than most APIs will have, but we're migrating to Sails from an older system and it's really feasible to cut that number drastically.

**Comment by mikermcneil** [comment]
@bradshja thanks for the heads up and for tracking the problem down.  The [commit you mentioned](https://github.com/balderdashy/sails/commit/1fdcc859eead5bbfa735f2887d0c7d3498832314) is probably an issue because somewhere those objects are getting _.cloneDeep'd  cc @sgress454

**Comment by sgress454** [comment]
@bradshja @mikermcneil interesting; I suppose we can make this optional as binding loaded modules' `module.exports.sails` to the Sails instance is a convenience for userland code, but not something that's relied on directly in Sails core.

**Comment by CWyrtzen** [comment]
Is this still going on?

**Comment by kevinburkeshyp** [comment]
fwiw i profiled this on my own system, which also loads a crap load of dependencies, and the vast, vast majority of the time was spent requiring every file, which is a sync operation. see https://kev.inburke.com/kevin/node-require-is-dog-slow/ for more info. might want to consider strategies to avoid loading so many files every time you want to run sails and/or keeping it running and only deleting/reloading code that you need to invalidate.

**Comment by bradshja** [comment]
We're currently using many fewer models while we continue to migrate from
our old API system to Sails, so right now it's much faster.  If we restore
the other 90% of the models then it might be much slower.

In our case, the vast majority of the time spent seemed to be in the route
binding, although that was just based on what was showing up in the logs
with Silly logging turned on.  It could well be that something else was
just making it look like the route binding was slow.

**Comment by sailsbot** [comment]
Thanks for posting, @bradshja.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by mikermcneil** [comment]
> In our case, the vast majority of the time spent seemed to be in the route
> binding, although that was just based on what was showing up in the logs
> with Silly logging turned on.

@bradshja  hmm... is blueprint routing enabled?

**Comment by odnarb** [comment]
Fwiw I have 58 controllers + 88 models. With blueprints enabled, it takes ~7-9 seconds to load on start.

With blueprints off it takes about half that time. It's a nice improvement. On load, does it wait for db connections to load up?

Still, applying 750+ models for an app sounds pretty insane..as in an edge case. :/
